libs/fileset/ignorer.go

```diff
@@ -17,3 +21,28 @@ func (nopIgnorer) IgnoreFile(path string) (bool, error) {
 func (nopIgnorer) IgnoreDirectory(path string) (bool, error) {
 	return false, nil
 }
+
+type includer struct {
+	matcher *ignore.GitIgnore
+}
+
+func newIncluder(includes []string) *includer {
+	matcher := ignore.CompileIgnoreLines(includes...)
+	return &includer{
+		matcher,
+	}
+}
+
+func (i *includer) IgnoreFile(path string) (bool, error) {
+	return i.ignore(path), nil
+}
+
+func (i *includer) IgnoreDirectory(path string) (bool, error) {
+	return false, nil
```

Comment:
To match `.gitignore` syntax, this needs to do the same as in `libs/git/view.go`.

If someone specifies `foo/bar/` (note the trailing `/`), it only matches the directory `bar` and not a file `bar`. The file system walker will then also skip walking the directory entirely, which can save a bunch of time.

Reply:
> If someone specifies foo/bar/ (note the trailing /), it only matches the directory bar and not a file bar. The file system walker will then also skip walking the directory entirely, which can save a bunch of time.

It indeed make sense for ignore logic. But for the include, the logic seems to be inversed: for the pattern `foo/bar/` we still need to look into the directory for file matches and can't really skip the dir altogether if there is no match, for example if we specify pattern like `foo/bar/*.go`. This pattern does not match directory but matches all the files inside of it. 

Or when we traverse the tree, we will try to call IgnoreDirectory on ".", "./foo", "./foo/bar" but use the pattern `foo/bar/*.go`. We need to make the matcher code more complex like generate different prefix patterns to be able to exclude code from traversing. 

Indeed, in this case lower complexity comes at the cost of performance

Reply:
Thanks, yeah this makes sense. Could you include this explanation in the code as well?

This is specific to the glob set _and_ the fact that we use it both for includes as well as excludes.